node/argstream.js

```diff
@@ -0,0 +1,276 @@
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+/*
+ * Provides federated streams for handling call arguments
+ *
+ * InArgStream is for handling incoming arg parts from call frames.  It handles
+ * dispatching the arg chunks into .arg{1,2,3} streams.
+ *
+ * OutArgStream is for creating outgoing arg parts by writing to .arg{1,2,3}
+ * streams.  It handles buffering as many parts as are written within one event
+ * loop tick into an Array of arg chunks.  Such array is then flushed using
+ * setImmediate.
+ *
+ * Due to the semantic complexity involved here, this code is tested by an
+ * accompanying exhaistive search test in test/argstream.js.  This test has
+ * both unit tests (disabled by default for speed) and an integration test.
+ */
+
+var inherits = require('util').inherits;
+var EventEmitter = require('events').EventEmitter;
+var PassThrough = require('stream').PassThrough;
+var Ready = require('ready-signal');
+var TypedError = require('error/typed');
+
+var ArgChunkOutOfOrderError = TypedError({
+    type: 'arg-chunk-out-of-order',
+    message: 'out of order arg chunk, current: {current} got: {got}',
+    current: null,
+    got: null
+});
+
+var ArgChunkGapError = TypedError({
+    type: 'arg-chunk-gap',
+    message: 'arg chunk gap, current: {current} got: {got}',
+    current: null,
+    got: null
+});
+
+function ArgStream() {
+    var self = this;
+    EventEmitter.call(self);
+    self.arg1 = StreamArg();
+    self.arg2 = StreamArg();
+    self.arg3 = StreamArg();
+
+    self.arg1.on('error', passError);
+    self.arg2.on('error', passError);
+    self.arg3.on('error', passError);
+    function passError(err) {
+        self.emit('error', err);
+    }
+
+    self.arg2.on('start', function onArg2Start() {
+        if (!self.arg1._writableState.ended) self.arg1.end();
+    });
+    self.arg3.on('start', function onArg3Start() {
+        if (!self.arg2._writableState.ended) self.arg2.end();
```

Comment:
Is this a programmer error ?

It seems like cutting arg2 mid stream when someone writes to arg3 is going to lead to corrupted data somewhere.